pandas/io/formats/xml.py

```diff
@@ -0,0 +1,583 @@
+        Raises
+        ------
+        LookupError
+            * If encoding is not available in codecs.
+        """
+
+        codecs.lookup(self.encoding)
+
+    def process_dataframe(self) -> Dict[Union[int, str], Dict[str, Any]]:
+        """
+        Adjust Data Frame to fit xml output.
+
+        This method will adjust underlying data frame for xml output,
+        including replacing missing entities and including indexes.
+        """
+
+        na_dict = {"None": self.na_rep, "NaN": self.na_rep, "nan": self.na_rep}
+
+        df = (
+            (self.fmt.frame.reset_index().applymap(str).replace(na_dict))
+            if self.index
+            else self.fmt.frame.applymap(str).replace(na_dict)
+        )
+
+        return df.to_dict(orient="index")
+
+    def handle_indexes(self) -> None:
+        """
+        Handle indexes.
+
+        This method will add indexes into attr_cols or elem_cols.
+        """
+
+        indexes: List[str] = [
+            x for x in self.frame_dicts[0].keys() if x not in self.orig_cols
+        ]
+
+        if self.attr_cols and self.index:
+            self.attr_cols = indexes + self.attr_cols
+
+        if self.elem_cols and self.index:
+            self.elem_cols = indexes + self.elem_cols
+
+    def get_prefix_uri(self) -> str:
+        """
+        Get uri of namespace prefix.
+
+        This method retrieves corresponding URI to prefix in namespaces.
+
+        Raises
+        ------
+        KeyError
+            *If prefix is not included in namespace dict.
+        """
+
+        raise AbstractMethodError(self)
+
+    def other_namespaces(self) -> dict:
+        """
+        Define other namespaces.
+
+        This method will build dictionary of namespaces attributes
+        for root element, conditionally with optional namespaces and
+        prefix.
+        """
+
+        nmsp_dict: Dict[str, str] = {}
+        if self.namespaces and self.prefix is None:
+            nmsp_dict = {"xmlns": n for p, n in self.namespaces.items() if p != ""}
+
+        if self.namespaces and self.prefix:
+            nmsp_dict = {"xmlns": n for p, n in self.namespaces.items() if p == ""}
+
+        return nmsp_dict
+
+    def build_attribs(self) -> None:
+        """
+        Create attributes of row.
+
+        This method adds attributes using attr_cols to row element and
+        works with tuples for multindex or hierarchical columns.
+        """
+
+        raise AbstractMethodError(self)
+
+    def build_elems(self) -> None:
+        """
+        Create child elements of row.
+
+        This method adds child elements using elem_cols to row element and
+        works with tuples for multindex or hierarchical columns.
+        """
+
+        raise AbstractMethodError(self)
+
+    def write_output(self) -> Optional[str]:
+        xml_doc = self.build_tree()
+        out_str: Optional[str] = xml_doc.decode(self.encoding).rstrip()
+
+        if self.path_or_buffer and isinstance(self.path_or_buffer, str):
```

Comment:
1. Can it be a `pathlib.Path`? `isinstance(stringify_path(self.path_or_buffer), str)`
2. `path_or_buffer: Optional[FilePathOrBuffer[str]]` but buffers do not seem to be handled (`get_handle`?) `FilePathOrBuffer[bytes]`?

Reply:
I assume this is already handled by `convert_io`?

Reply:
I cannot pass `mypy` with either approach for `open()`. See message:

> Argument 1 to "open" has incompatible type "Union[_PathLike[str], str, IO[str], RawIOBase, BufferedIOBase, TextIOBase, mmap]"; expected "Union[Union[str, bytes, _PathLike[str], _PathLike[bytes]], int]"  [arg-type]

I recall running into this situation earlier. Also, I did try `get_handle` from pandas.io.common but ran into `TypeError` for sending bytes and not str. The method will need adjustment for xml content.

Reply:
`get_handle` can deal with bytes: `get_handle(..., is_text=False)`.

You could call `stringify_path` in init to save `FileOrBuffer` (without the `Path`)? I didn't try whether mypy complains :)

Reply:
THAT was what I needed! I can now implement `get_handle()` in `to_xml` and `read_xml` which supports `compression` and `storage_options`. Now to add needed pytests. Thanks!